### ISSUE TITLE
critical CVEs showed in container vulnerability scan with nats:2.9 image

### ISSUE THREAD

**Issue by jzhoucliqr** [nats-io/nats-server#3992]
Hi there, 

We've seen some weird output when run vulnerability scan on the nats:2.9 container image. We do not see such behavior in nats:2.8 and earlier. Does one have any idea what could've caused this? Thanks! 

when run with Twistlock, reports like below; 

```
root@ubuntuguest:~# docker pull nats:2.9 && ./tw.sh nats:2.9
2.9: Pulling from library/nats
Digest: sha256:58e483681983f87fdc738e980d582d4cc7218914c4f0056d36bab0c5acfc5a8b
Status: Downloaded newer image for nats:2.9
docker.io/library/nats:2.9

Scan results for: image nats:2.9 sha256:74665a2673c052e85d3417cec864d1aabab3263f3328051037a133d9b85765ce
Vulnerabilities
+---------------------+----------+------+------------------------------------------+---------+-------------------------+-----------+------------+------------------------------------------------------+
|         CVE         | SEVERITY | CVSS |                 PACKAGE                  | VERSION |         STATUS          | PUBLISHED | DISCOVERED |                     DESCRIPTION                      |
+---------------------+----------+------+------------------------------------------+---------+-------------------------+-----------+------------+------------------------------------------------------+
| GHSA-j756-f273-xhp4 | critical | 9.00 | github.com/nats-io/nats-server/v2        | (devel) | fixed in 2.2.0          | > 1 years | < 1 hour   | (This advisory is canonically                        |
|                     |          |      |                                          |         | > 1 years ago           |           |            | <https://advisories.nats.io/CVE/CVE-2021-3127.txt>)  |
|                     |          |      |                                          |         |                         |           |            |  ## Problem Description  The NATS server provides    |
|                     |          |      |                                          |         |                         |           |            | for Subjects which...                                |
+---------------------+----------+------+------------------------------------------+---------+-------------------------+-----------+------------+------------------------------------------------------+
| CVE-2020-26892      | critical | 9.00 | github.com/nats-io/nats-server/v2        | (devel) | fixed in 2.1.9          | > 2 years | < 1 hour   | The JWT library in NATS nats-server before 2.1.9     |
|                     |          |      |                                          |         | > 2 years ago           |           |            | has Incorrect Access Control because of how          |
|                     |          |      |                                          |         |                         |           |            | expired credentials are handled.                     |
+---------------------+----------+------+------------------------------------------+---------+-------------------------+-----------+------------+------------------------------------------------------+
| GHSA-2c64-vj8g-vwrq | high     | 7.00 | github.com/nats-io/nats-server/v2        | (devel) | fixed in 2.1.9          | > 1 years | < 1 hour   | (This advisory is canonically                        |
|                     |          |      |                                          |         | > 1 years ago           |           |            | https://advisories.nats.io/CVE/CVE-2020-26892.txt    |
|                     |          |      |                                          |         |                         |           |            | )  ## Problem Description  NATS nats-server          |
|                     |          |      |                                          |         |                         |           |            | through 2020-10-07 has Inc...                        |
+---------------------+----------+------+------------------------------------------+---------+-------------------------+-----------+------------+------------------------------------------------------+
| CVE-2021-3127       | high     | 7.00 | github.com/nats-io/nats-server/v2/server | (devel) | fixed in 2.2.0          | > 2 years | < 1 hour   | NATS Server 2.x before 2.2.0 and JWT library         |
|                     |          |      |                                          |         | > 2 years ago           |           |            | before 2.0.1 have Incorrect Access Control because   |
|                     |          |      |                                          |         |                         |           |            | Import Token bindings are mishandled.                |
+---------------------+----------+------+------------------------------------------+---------+-------------------------+-----------+------------+------------------------------------------------------+
| CVE-2020-28466      | high     | 7.00 | github.com/nats-io/nats-server/v2/server | (devel) | fixed in 2.2.0          | > 2 years | < 1 hour   | This affects all versions of package                 |
|                     |          |      |                                          |         | > 2 years ago           |           |            | github.com/nats-io/nats-server/server. Untrusted     |
|                     |          |      |                                          |         |                         |           |            | accounts are able to crash the server using          |
|                     |          |      |                                          |         |                         |           |            | configs that represe...                              |
+---------------------+----------+------+------------------------------------------+---------+-------------------------+-----------+------------+------------------------------------------------------+
| CVE-2020-26521      | high     | 7.00 | github.com/nats-io/nats-server/v2        | (devel) | fixed in 2.1.9          | > 2 years | < 1 hour   | The JWT library in NATS nats-server before 2.1.9     |
|                     |          |      |                                          |         | > 2 years ago           |           |            | allows a denial of service (a nil dereference in     |
|                     |          |      |                                          |         |                         |           |            | Go code).                                            |
```

When run with trivy, it reported version error then doesn't give any output:

```
root@ubuntuguest:~/code/nats-server# trivy image nats:2.9
2023-03-24T17:47:58.829Z	INFO	Vulnerability scanning is enabled
2023-03-24T17:47:58.829Z	INFO	Secret scanning is enabled
2023-03-24T17:47:58.829Z	INFO	If your scanning is slow, please try '--scanners vuln' to disable secret scanning
2023-03-24T17:47:58.829Z	INFO	Please see also https://aquasecurity.github.io/trivy/v0.37/docs/secret/scanning/#recommendation for faster secret detection
2023-03-24T17:47:58.843Z	INFO	Number of language-specific files: 1
2023-03-24T17:47:58.843Z	INFO	Detecting gobinary vulnerabilities...
2023-03-24T17:47:58.843Z	WARN	version error ((devel)): malformed version: (devel)
2023-03-24T17:47:58.843Z	WARN	version error ((devel)): malformed version: (devel)
2023-03-24T17:47:58.843Z	WARN	version error ((devel)): malformed version: (devel)
2023-03-24T17:47:58.843Z	WARN	version error ((devel)): malformed version: (devel)
2023-03-24T17:47:58.843Z	WARN	version error ((devel)): malformed version: (devel)
2023-03-24T17:47:58.844Z	WARN	version error ((devel)): malformed version: (devel)
2023-03-24T17:47:58.844Z	WARN	version error ((devel)): malformed version: (devel)
2023-03-24T17:47:58.844Z	WARN	version error ((devel)): malformed version: (devel)
2023-03-24T17:47:58.844Z	WARN	version error ((devel)): malformed version: (devel)
2023-03-24T17:47:58.844Z	WARN	version error ((devel)): malformed version: (devel)
2023-03-24T17:47:58.844Z	WARN	version error ((devel)): malformed version: (devel)
```

**Comment by wallyqs** [comment]
Investigating why the malformed version string is being reported since can be confusing, but with trivy you can use an exit code to determine whether there were issues or not in the scan.  From the trivy report for 2.9.15 looks like there are no issues (but not sure about the twistlock output or what is it checking atm 🤔 ):

```sh
$ trivy image nats:2.9.15 --exit-code 1
echo $?
0

$ trivy image nats:2.8.0 --exit-code 1
echo $?
1
```

**Comment by wallyqs** [comment]
The grype output seems similar to that of twistlock, removing `devel` should fix these too:

```
 grype nats:2.9.15
 ✔ Vulnerability DB        [no update available]
 ✔ Parsed image            
 ✔ Cataloged packages      [11 packages]
 ✔ Scanning image...       [6 vulnerabilities]
   ├── 2 critical, 2 high, 0 medium, 2 low, 0 negligible
   └── 6 fixed

NAME                               INSTALLED  FIXED-IN  TYPE       VULNERABILITY        SEVERITY 
github.com/nats-io/nats-server/v2  (devel)    2.1.9     go-module  GHSA-2c64-vj8g-vwrq  High      
github.com/nats-io/nats-server/v2  (devel)    2.1.9     go-module  GHSA-4w5x-x539-ppf5  Critical  
github.com/nats-io/nats-server/v2  (devel)    2.1.9     go-module  GHSA-h2fg-54x9-5qhq  High      
github.com/nats-io/nats-server/v2  (devel)    2.1.9     go-module  GHSA-hmm9-r2m2-qg9w  Low       
github.com/nats-io/nats-server/v2  (devel)    2.2.0     go-module  GHSA-gwj5-3vfq-q992  Low       
github.com/nats-io/nats-server/v2  (devel)    2.2.0     go-module  GHSA-j756-f273-xhp4  Critical  
```

**Comment by wallyqs** [comment]
Adding the vcsinfo as in https://github.com/nats-io/nats-server/pull/3995 fixed it for nightly:

Before:

```
$ grype synadia/nats-server:nightly-20230326
 ✔ Vulnerability DB        [no update available]
 ✔ Parsed image            
 ✔ Cataloged packages      [113 packages]
 ✔ Scanning image...       [10 vulnerabilities]
   ├── 2 critical, 3 high, 1 medium, 2 low, 0 negligible (2 unknown)
   └── 8 fixed
NAME                               INSTALLED  FIXED-IN  TYPE       VULNERABILITY        SEVERITY 
github.com/nats-io/nats-server/v2  (devel)    2.1.9     go-module  GHSA-2c64-vj8g-vwrq  High      
github.com/nats-io/nats-server/v2  (devel)    2.1.9     go-module  GHSA-4w5x-x539-ppf5  Critical  
github.com/nats-io/nats-server/v2  (devel)    2.1.9     go-module  GHSA-h2fg-54x9-5qhq  High      
github.com/nats-io/nats-server/v2  (devel)    2.1.9     go-module  GHSA-hmm9-r2m2-qg9w  Low       
github.com/nats-io/nats-server/v2  (devel)    2.2.0     go-module  GHSA-gwj5-3vfq-q992  Low       
github.com/nats-io/nats-server/v2  (devel)    2.2.0     go-module  GHSA-j756-f273-xhp4  Critical  
libcrypto3                         3.0.8-r0   3.0.8-r1  apk        CVE-2023-0464        Unknown   
libssl3                            3.0.8-r0   3.0.8-r1  apk        CVE-2023-0464        Unknown   

$ grype synadia/nats-server:nightly-20230327
 ✔ Vulnerability DB        [no update available]
 ✔ Parsed image            
 ✔ Cataloged packages      [113 packages]
 ✔ Scanning image...       [10 vulnerabilities]
   ├── 2 critical, 3 high, 1 medium, 2 low, 0 negligible (2 unknown)
   └── 8 fixed

NAME                               INSTALLED  FIXED-IN  TYPE       VULNERABILITY        SEVERITY 
github.com/nats-io/nats-server/v2  (devel)    2.1.9     go-module  GHSA-2c64-vj8g-vwrq  High      
github.com/nats-io/nats-server/v2  (devel)    2.1.9     go-module  GHSA-4w5x-x539-ppf5  Critical  
github.com/nats-io/nats-server/v2  (devel)    2.1.9     go-module  GHSA-h2fg-54x9-5qhq  High      
github.com/nats-io/nats-server/v2  (devel)    2.1.9     go-module  GHSA-hmm9-r2m2-qg9w  Low       
github.com/nats-io/nats-server/v2  (devel)    2.2.0     go-module  GHSA-gwj5-3vfq-q992  Low       
github.com/nats-io/nats-server/v2  (devel)    2.2.0     go-module  GHSA-j756-f273-xhp4  Critical  
libcrypto3                         3.0.8-r0   3.0.8-r1  apk        CVE-2023-0464        Unknown   
libssl3                            3.0.8-r0   3.0.8-r1  apk        CVE-2023-0464        Unknown   
```

After merging https://github.com/nats-io/nats-server/pull/3995 : 

```sh
grype synadia/nats-server:nightly-20230328
 ✔ Vulnerability DB        [no update available]
 ✔ Parsed image            
 ✔ Cataloged packages      [112 packages]
 ✔ Scanning image...       [4 vulnerabilities]
   ├── 0 critical, 1 high, 1 medium, 0 low, 0 negligible (2 unknown)
   └── 2 fixed

NAME        INSTALLED  FIXED-IN  TYPE  VULNERABILITY  SEVERITY 
libcrypto3  3.0.8-r0   3.0.8-r1  apk   CVE-2023-0464  Unknown   
libssl3     3.0.8-r0   3.0.8-r1  apk   CVE-2023-0464  Unknown  
```

With trivy, before:

```
trivy image synadia/nats-server:nightly-20230327
2023-03-27T23:35:30.032-0700	INFO	Vulnerability scanning is enabled
2023-03-27T23:35:30.032-0700	INFO	Secret scanning is enabled
2023-03-27T23:35:30.032-0700	INFO	If your scanning is slow, please try '--scanners vuln' to disable secret scanning
2023-03-27T23:35:30.032-0700	INFO	Please see also https://aquasecurity.github.io/trivy/v0.38/docs/secret/scanning/#recommendation for faster secret detection
2023-03-27T23:35:32.422-0700	INFO	Detected OS: alpine
2023-03-27T23:35:32.422-0700	INFO	Detecting Alpine vulnerabilities...
2023-03-27T23:35:32.422-0700	INFO	Number of language-specific files: 3
2023-03-27T23:35:32.422-0700	INFO	Detecting gobinary vulnerabilities...
2023-03-27T23:35:32.422-0700	WARN	version error ((devel)): malformed version: (devel)
2023-03-27T23:35:32.422-0700	WARN	version error ((devel)): malformed version: (devel)
2023-03-27T23:35:32.422-0700	WARN	version error ((devel)): malformed version: (devel)
2023-03-27T23:35:32.422-0700	WARN	version error ((devel)): malformed version: (devel)
2023-03-27T23:35:32.422-0700	WARN	version error ((devel)): malformed version: (devel)
2023-03-27T23:35:32.422-0700	WARN	version error ((devel)): malformed version: (devel)
2023-03-27T23:35:32.422-0700	WARN	version error ((devel)): malformed version: (devel)
2023-03-27T23:35:32.422-0700	WARN	version error ((devel)): malformed version: (devel)
2023-03-27T23:35:32.422-0700	WARN	version error ((devel)): malformed version: (devel)
2023-03-27T23:35:32.422-0700	WARN	version error ((devel)): malformed version: (devel)
2023-03-27T23:35:32.422-0700	WARN	version error ((devel)): malformed version: (devel)
```

After merging, no more warnings:

```
trivy image synadia/nats-server:nightly-20230328
2023-03-27T23:35:43.689-0700	INFO	Vulnerability scanning is enabled
2023-03-27T23:35:43.689-0700	INFO	Secret scanning is enabled
2023-03-27T23:35:43.689-0700	INFO	If your scanning is slow, please try '--scanners vuln' to disable secret scanning
2023-03-27T23:35:43.689-0700	INFO	Please see also https://aquasecurity.github.io/trivy/v0.38/docs/secret/scanning/#recommendation for faster secret detection
2023-03-27T23:35:45.080-0700	INFO	Detected OS: alpine
2023-03-27T23:35:45.080-0700	INFO	Detecting Alpine vulnerabilities...
2023-03-27T23:35:45.082-0700	INFO	Number of language-specific files: 3
2023-03-27T23:35:45.082-0700	INFO	Detecting gobinary vulnerabilities...
```

**Comment by wallyqs** [comment]
Thanks for the report @jzhoucliqr this has been fixed for our nightly releases already and will be fixed in the v2.9.16 release as well.
